---
title: "Differential Expression Analysis - FUBP1 KD"
subtitle: "Differential Expression Analysis of FUBP1 KD and WT cell lines"
author: "Martin Brand"
date: "`r format(Sys.time(), '%B %e, %Y')`"
output:
  bookdown::html_document2:
    code_folding: hide
    fig_caption: yes
    number_sections: no
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
    toc_collapsed: yes
    fig_width: 8 
    fig_height: 4 
  bookdown::pdf_document2:
    keep_tex: yes
    fig_caption: yes
    number_sections: no
    toc: yes
    toc_depth: 3
graphics: yes
header-includes:
- \makeatletter\renewcommand*{\fps@figure}{h}\makeatother
- \usepackage{placeins}
geometry: margin=1in
fontsize: 18pt
bibliography: references.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Scientific Background and Research Question
Scientific background of this analysis is the research into the FUBP1 protein. FUBP1 has been identified as a DNA-binding protein by @fubp1_dna. In recent years, studies have shown that FUBP1 is also an RNA-binding protein that plays an important role in alternative splicing by @fubp1_rna. However, the exact function of FUBP1 is not yet known. In this project, I want to explore the genetic dependencies of FUBP1. The aim of this analysis was to find out whether knock-down of the FUBP1 gene leads to a change in gene expression of the U2-related genes published by @wahl.

```{r, message=FALSE, warning=FALSE}
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(apeglm)
library(ggplot2)
library(ggrastr)
library(dplyr)
library(ggrepel)
library(reshape2)
```

# Preprocessing of the Data
I loaded the read counts tables which output from featureCounts and created an experiment design table with two wild type cell lines and two cell lines with FUBP1 KD.
```{r}
dir <- "C:/Users/Martin/Documents/Master/Forschungspraktikum/differential_expression_analysis/project/data/"
files = list.files(paste0(dir,"feature_counts_output"))

sampleFiles <- files
sampleCondition <- factor(rep(c("Ctrl", "KD"), each=2))
sampleReplicate <- factor(paste("Rep", 1:2, sep = "_",2))

sampleTable <- data.frame(sampleName = sampleFiles,
                          fileName = sampleFiles,
                          condition = sampleCondition,
                          replicate = sampleReplicate)
sampleTable
```


I put the data into a specific format that is required for the DESeq analysis and performed the analysis. The result from the DESeq analysis contains the number of read counts for genes for each sample.
```{r, message=FALSE}
# fit format
count_table <- read.table(paste0(dir, "feature_counts_output/", files[1]), header = T) 
count_table <- count_table[, c(1,ncol(count_table))]

for (i in 2:4) {
  tmp <- read.table(paste0(dir, "feature_counts_output/", files[i]), header = T)
  tmp <- tmp[, c(1,ncol(tmp))]
  count_table <- merge(count_table, tmp, by = "Geneid")
}

colnames(count_table)[2:5] <- c("WT1", "WT2", "KD1", "KD2")

# problem: doubled genes => delete all entries with zero counts
count_table <- count_table[rowSums(count_table[,2:5]) != 0,]

# delete version number of gene IDs
count_table$Geneid <- substr(count_table$Geneid, 1, 15)

# move gene IDs from column to row names
rownames(count_table) <- count_table$Geneid
count_table$Geneid <- NULL

# analysis
dds <- DESeqDataSetFromMatrix(countData = count_table,
                              colData = sampleTable,
                              design = ~ condition)

dds <- DESeq(dds)

head(counts(dds))
```

# Quality Control
For quality control, I calculated the differences between the samples using the regulized log transformation and visualized the results in a heatmap. As the heatmap shows, the differences between WT and FUBP1 KD cell lines are larger than the differences between replicates.
```{r}
rld <- rlog(dds, blind = FALSE)
sampleDists <- dist(t(assay(rld)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rld$condition, rld$replicate, sep = "-")
colnames(sampleDistMatrix) <- rownames(sampleDistMatrix)
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)

pheatmap(sampleDistMatrix, clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, col=colors, legend = FALSE)
```

The second quality check is done with PCA. The FUBP1 KD and WT cell lines show a significant difference in PC1, which accounts for 97.7% of the variance in the data.
```{r}
pcaData <- plotPCA(rld, intgroup=c("condition", "replicate"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData,"percentVar"), digits=1)
ggplot(pcaData, aes(x=PC1, y=PC2, color=condition, shape=replicate)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  theme(aspect.ratio = 1)
```
Considering both quality checks, we can say that the data quality is very good as we can see the expected patterns. Gene expression for FUBP1 KD and WT cell lines differs more than for replicates.

# Up and Downregulated Genes
To identify the regulated genes upon the FUBP1 KD, I extracted the result table, which contains the log2 fold changes upon the treatment and the adjusted p-values. Besides, I also adjusted the log2 fold changes based on the read counts using the package from @apeglm.
```{r, message=FALSE}
# extract result table and add column for adjusted log fold changes
res <- results(dds, contrast = c("condition","KD","Ctrl"))
res.shrink <- lfcShrink(dds = dds, res = res, coef = 2)
res$log2FoldChange.shrink <- res.shrink$log2FoldChange
res <- as.data.frame(res)
# print only important columns
head(res[c(1,2,6)])
```

For quality control of the test result, I visualized the distribution of adjusted p-values. The distribution looks as expected, because the left bar shows that there are changes in gene expression. 
```{r}
hist(res$padj, main = "Distribution of adjusted P-Values", xlab = "Adjusted P-Value")
```

I visualized the differential expression results in a MA plot in which every data point represents a gene. The y-axis shows the log2 fold change, which indicates gene expression changes upon the FUBP1 KD. The x-axis shows the log transformed mean value of normalized counts (assigned reads), which is an indicator of the reliability of the data. All data points marked as blue represent genes with a p-value greater than 0.1. U2-related genes are labeled with their gene names.
```{r}
res$regulated <- res$padj < 0.1
res$regulated[is.na(res$regulated)] <- FALSE
res <- mutate(res, up_down = case_when(
  regulated & log2FoldChange > 0 ~ "up",
  regulated & log2FoldChange < 0 ~ "down",
  TRUE ~ "not"
))
res$up_down <- factor(res$up_down)

# mark U2 related genes for labels
labeled_genes <- c(FUBP1="ENSG00000162613", U2AF2="ENSG00000063244", SF1="ENSG00000168066",
                   PUF60="ENSG00000179950", SPF30="ENSG00000119953", SPF31="ENSG00000126698",
                   SPF45="ENSG00000134453", CHERP="ENSG00000085872", SR140="ENSG00000241537",
                   HPrp43="ENSG00000109606", HPrp5="ENSG00000145833")
index <- match(labeled_genes, rownames(res))

# add column for labeling chosen genes
res$to.label <- ""
res$to.label[index] <- names(labeled_genes)

ggplot(res, aes(x=log10(baseMean), y=log2FoldChange, col=up_down, label=to.label))+
  geom_point(size=0.7) +
  geom_point(data=subset(res,to.label!=""),color="black") +
  geom_label_repel(max.overlaps = Inf, show.legend = FALSE)
```

In the plot above, there are many genes with high absolute log2 fold changes and low read counts. These data points are not reliable. That is why the log2 fold changes have been adjusted to the read counts. The plot below shows the adjusted log2 fold changes. It can be seen that the number of data points with low reliability and high log2 fold changes could be significantly reduced. This result suggests the KD of FUBP1 does not lead to significant change of expression of U2-related genes.
```{r}
ggplot(res, aes(x=log10(baseMean), y=log2FoldChange.shrink, col=up_down, label=to.label))+
  geom_point(size=0.7) +
  geom_point(data=subset(res,to.label!=""),color="black") +
  geom_label_repel(max.overlaps = Inf, show.legend = FALSE)
```
I visualized the normalized read counts of FUBP1 and U2-related genes in the different samples. The number of read counts differs strongly for the FUBP1 gene but very weakly or not for all U2-related genes.
```{r, message=FALSE}
counts <- counts(dds, normalized=TRUE)
colnames(counts) <- colData(dds)$condition
counts <- melt(counts) # rows all columns under each other
colnames(counts) <- c("gene.ID", "condition", "norm.count")
counts.labeled_genes <- subset(counts, gene.ID %in% labeled_genes)

labeled_genes.subst = names(labeled_genes)
names(labeled_genes.subst) <- labeled_genes

ggplot(counts.labeled_genes, aes(x=condition, y=norm.count)) + 
  geom_point() +
  scale_y_log10() +
  facet_wrap(~gene.ID, scales="free", labeller = as_labeller(labeled_genes.subst)) +
  ylim(0, 7000) +
  ylab("Normalized Counts")
```

# Summary and Outlook
In summary, knock-down of the FUBP1 gene does not lead to a change in expression of U2-related genes. Further analysis could done using the approach by @chan_wrn.