---
title: "Differential Expression Analysis"
subtitle: "Differential Expression Analysis after KD of FUBP1 gene"
author: "Martin Brand"
date: "2022-10-08"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message = FALSE, results="hide", echo=TRUE}
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(apeglm)
library(ggplot2)
library(ggrastr)
library(dplyr)
library(ggrepel)
library(reshape2)
```

We specify the path to the directory with the data from featureCount and save
the corresponding files.
```{r, message = FALSE, results="hide", echo=TRUE}
dir <- "C:/Users/Martin/Documents/Master/Forschungspraktikum/differential_expression_analysis/project/data/"
files = list.files(paste0(dir,"feature_counts_output"))
```

## DESeq Analysis
We create a sample table with two wild type cell lines and two cell lines with
FUBP1 KD.
```{r, message = FALSE, echo=TRUE}
sampleFiles <- files
sampleCondition <- factor(rep(c("Ctrl", "KD"), each=2))
sampleReplicate <- factor(paste("Rep", 1:2, sep = "_",2))

sampleTable <- data.frame(sampleName = sampleFiles,
                          fileName = sampleFiles,
                          condition = sampleCondition,
                          replicate = sampleReplicate)
sampleTable
```


We create a count table that we need for the function 
**DESeqDataSetFromMatrix()**. The count table must contain the number of assigned
reads for each gene in each cell line. **DESeqDataSetFromMatrix()** returns the
**DESeqDataSet** that we need for further analysis. After running the analysis
we look at the final count table for a better understanding.

```{r, message = FALSE, warning=TRUE, echo=TRUE}
df_count <- read.table(paste0(dir, "feature_counts_output/", files[1]), header = T) 
df_count <- df_count[, c(1,ncol(df_count))]

for (i in 2:4) {
  tmp <- read.table(paste0(dir, "feature_counts_output/", files[i]), header = T)
  tmp <- tmp[, c(1,ncol(tmp))]
  df_count <- merge(df_count, tmp, by = "Geneid")
}

colnames(df_count)[2:5] <- c("WT1", "WT2", "KD1", "KD2")
# delete all entries with zero counts
# problem: doubled genes
df_count <- df_count[rowSums(df_count[,2:5]) != 0,]
# delete version number of gene IDs
gene.IDs <- substr(df_count$Geneid, 1, 15)
# move gene IDs from column to row names
rownames(df_count) <- gene.IDs 
df_count$Geneid <- NULL

dds <- DESeqDataSetFromMatrix(countData = df_count,
                              colData = sampleTable,
                              design = ~ condition)

dds <- DESeq(dds)

head(counts(dds))
```

## Heatmap
For quality control we present the results in a heatmap. We see that the differences between wild types and FUBP1 KD cell lines are larger than between the two wild types or the two FUBP1 KD cell lines.
```{r, message = FALSE, results="hide", echo=TRUE}
rld <- rlog(dds, blind = FALSE)
sampleDists <- dist(t(assay(rld)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rld$condition, rld$replicate, sep = "-")
colnames(sampleDistMatrix) <- rownames(sampleDistMatrix)
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)

pheatmap(sampleDistMatrix, clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, col=colors)
```

## Principal Component Analysis
We do a PCA to clarify the differences between the control cell lines and treated cell lines. The difference of PC1 is strikingly large. 
```{r, message = FALSE, echo=TRUE}
pcaData <- plotPCA(rld, intgroup=c("condition", "replicate"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData,"percentVar"), digits=1)
ggplot(pcaData, aes(x=PC1, y=PC2, color=condition, shape=replicate)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  theme(aspect.ratio = 1)
```

## Find most up and downregulated genes
We format the data for further calculations and have a look at the new format.
```{r, message = FALSE, echo=TRUE}
res <- results(dds, contrast = c("condition","KD","Ctrl"))
head(res[c(2,5)])
```

The log2FoldChange is given for every gene. We use this parameter to find the
genes whose expression has changed the most.
```{r, message = FALSE, echo=TRUE}
TopUP <- res[order(res$log2FoldChange, decreasing = TRUE),c(2,5)]
TopUP <- TopUP[1:5,]
TopUP
```

We can save the results and retrieve them at any time.
```{r, message = FALSE, results="hide", echo=TRUE}
write.table(TopUP, file = paste0(dir,"DESeq2_up.tab"), quote=FALSE,
            sep="\t")
TopUP <- read.table(paste0(dir,"DESeq2_up.tab"))
```

## MA plots
We visualize the results as an MA plot. Every dot represents a gene. The y axis shows the logfoldchange which indicates how much the expression of each gene has changed. The x axis shows the mean of normalized counts (assigned reads) which is an indicator of the reliability of the data. All genes with a P-value greater than 0.1 are defined as regulated and marked with blue.
```{r, message = FALSE, echo=TRUE}
plotMA(res, main="MA plot")
```

We remove the noise (high logfoldchanges by low read counts) using the function lfcShrink. The logfoldchanges are adapted to the read counts. We save the shrunked logfoldchanges in our result table. We see in the figure that we have no more outliers with low reliability. 
```{r, message = FALSE, echo=TRUE}
res.shrink <- lfcShrink(dds = dds, res = res, coef = 2)
res$log2FoldChange.shrink <- res.shrink$log2FoldChange
plotMA(res.shrink, main="MA plot with shrunken fold changes")
```

We use ggplot to reproduce the MA plot with shrunken fold changes because it allows us to customise like highlighting genes, e.g. FUBP1. We see that the gene was clearly regulated down. The knockdown of FUBP1 seems to have little or no effect on the expression of most U2 related genes. Only SPF31 and HPrp43 are downregulated.
```{r, message = FALSE, warning= TRUE, echo=TRUE}
res.df <- as.data.frame(res)
# add column which tells about regulation
res.df$regulated <- res.df$padj < 0.01
res.df$regulated[is.na(res.df$regulated)] <- FALSE
# add columns which tells about up and down regulation
res.df <- mutate(res.df, up_down = case_when(
  regulated & log2FoldChange > 0 ~ "up",
  regulated & log2FoldChange < 0 ~ "down",
  TRUE ~ "not"
))
res.df$up_down <- factor(res.df$up_down)

# mark genes for labels
labeled_genes <- c(FUBP1="ENSG00000162613", U2AF2="ENSG00000063244", SF1="ENSG00000168066",
                   PUF60="ENSG00000179950", SPF30="ENSG00000119953", SPF31="ENSG00000126698",
                   SPF45="ENSG00000134453", CHERP="ENSG00000085872", SR140="ENSG00000241537",
                   HPrp43="ENSG00000109606", HPrp5="ENSG00000145833")
index <- match(labeled_genes, rownames(res))

# add column for labeling chosen genes
res.df$to.label <- NA
res.df$to.label[index] <- names(labeled_genes)

# ggplot(res.df, aes(x=log10(baseMean), y=log2FoldChange.shrink, col=up_down, label=to.label)) +
#   # ggrastr::rasterise(geom_point(), dpi=300) +
#   geom_point(res.df, is.na(to.label), color=res.df$color) +
#   geom_point(data=subset(res.df,is.na(to.label) & up_down=="up"),color="blue") +
#   geom_point(data=subset(res.df,is.na(to.label) & up_down=="down"),color="deepskyblue") +
#   geom_point(data=subset(res.df,is.na(to.label) & regulated==FALSE),color="darkgrey") +
#   geom_point(data=subset(res.df,!is.na(to.label)),color="red") +
#   geom_label_repel(max.overlaps = Inf) +
#   coord_cartesian(ylim = c(-5,6))

ggplot(res.df, aes(x=log10(baseMean), y=log2FoldChange.shrink, col=up_down, label=to.label))+
  geom_point() +
  geom_point(data=subset(res.df,!is.na(to.label)),color="black") +
  geom_label_repel(max.overlaps = Inf, show.legend = FALSE)


#problem: a in legend


```

We visualize the data as P value distribution. Most genes have a very low
P value. 
```{r, message = FALSE, echo=TRUE}
hist(res$pvalue)
```

We visualize the read counts of FUBP1 and U2 related genes in the different samples. We see that the number of read counts differs largely for the FUBP1 gene but very little or not for all U2 related genes.
```{r, message = FALSE, echo=TRUE}
counts <- counts(dds, normalized=TRUE)
colnames(counts) <- colData(dds)$condition
counts.df <- melt(counts) # rows all columns under each other
colnames(counts.df) <- c("gene.ID", "condition", "norm.count")
counts.labeled_genes <- subset(counts.df, gene.ID %in% labeled_genes)

# change data structure of labeled genes for labeller parameter
labeled_genes.subst = names(labeled_genes)
names(labeled_genes.subst) <- labeled_genes

ggplot(counts.labeled_genes, aes(x=condition, y=norm.count)) + 
  geom_point() +
  scale_y_log10() +
  facet_wrap(~gene.ID, scales="free", labeller = as_labeller(labeled_genes.subst)) +
  ylim(0, 7000)
```

## Saving data
Saving the data as .Rdata allows to efficiently load them into follow-up script.
```{r, message = FALSE, echo=TRUE}
res.df <- as.data.frame(res)
save(res.df, file=paste0(dir, "results_table.Rdata"))
```
