---
title: "Differential Expression Analysis"
subtitle: "Differential Expression Analysis in FUBP1 KD cell lines compared to WT cell lines"
author: "Martin Brand"
date: "`r format(Sys.time(), '%B %e, %Y')`"
output:
  bookdown::html_document2:
    code_folding: hide
    fig_caption: yes
    number_sections: no
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
    toc_collapsed: yes
    fig_width: 8 
    fig_height: 4 
  bookdown::pdf_document2:
    keep_tex: yes
    fig_caption: yes
    number_sections: no
    toc: yes
    toc_depth: 3
graphics: yes
header-includes:
- \makeatletter\renewcommand*{\fps@figure}{h}\makeatother
- \usepackage{placeins}
geometry: margin=1in
fontsize: 18pt
bibliography: references.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message=FALSE}
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(apeglm)
library(ggplot2)
library(ggrastr)
library(dplyr)
library(ggrepel)
library(reshape2)
```

## Preprocessing of the Data
I load the raw files that contain the read counts and create a sample table with two wild type cell lines and two cell lines with FUBP1 KD.
```{r}
dir <- "C:/Users/Martin/Documents/Master/Forschungspraktikum/differential_expression_analysis/project/data/"
files = list.files(paste0(dir,"feature_counts_output"))

sampleFiles <- files
sampleCondition <- factor(rep(c("Ctrl", "KD"), each=2))
sampleReplicate <- factor(paste("Rep", 1:2, sep = "_",2))

sampleTable <- data.frame(sampleName = sampleFiles,
                          fileName = sampleFiles,
                          condition = sampleCondition,
                          replicate = sampleReplicate)
sampleTable
```


I put the data into a specific format that is required for the DESeq analysis and perform the analysis. The result table shows the number of read counts for every gene.
```{r, message=FALSE}
# fit format
count_table <- read.table(paste0(dir, "feature_counts_output/", files[1]), header = T) 
count_table <- count_table[, c(1,ncol(count_table))]

for (i in 2:4) {
  tmp <- read.table(paste0(dir, "feature_counts_output/", files[i]), header = T)
  tmp <- tmp[, c(1,ncol(tmp))]
  count_table <- merge(count_table, tmp, by = "Geneid")
}

colnames(count_table)[2:5] <- c("WT1", "WT2", "KD1", "KD2")

# problem: doubled genes => delete all entries with zero counts
count_table <- count_table[rowSums(count_table[,2:5]) != 0,]

# delete version number of gene IDs
count_table$Geneid <- substr(count_table$Geneid, 1, 15)

# move gene IDs from column to row names
rownames(count_table) <- count_table$Geneid
count_table$Geneid <- NULL

# analysis
dds <- DESeqDataSetFromMatrix(countData = count_table,
                              colData = sampleTable,
                              design = ~ condition)

dds <- DESeq(dds)

head(counts(dds))
```

## Quality Control
For quality control, I calculate the differences between the samples using the regulised log transformation and visualize the results in a heatmap. The differences of the read counts between WT cell lines and FUBP1 KD cell lines are larger than between the two WT cell lines or the two FUBP1 KD cell lines.
```{r}
rld <- rlog(dds, blind = FALSE)
sampleDists <- dist(t(assay(rld)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rld$condition, rld$replicate, sep = "-")
colnames(sampleDistMatrix) <- rownames(sampleDistMatrix)
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)

pheatmap(sampleDistMatrix, clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, col=colors, legend = FALSE)
```

A second quality check is done with PCA. The result shows differences between the control cell lines and the FUBP1 KD cell lines. PC1 is responsible for 97.7% of the variance and shows a large difference between the groups.
```{r}
pcaData <- plotPCA(rld, intgroup=c("condition", "replicate"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData,"percentVar"), digits=1)
ggplot(pcaData, aes(x=PC1, y=PC2, color=condition, shape=replicate)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  theme(aspect.ratio = 1)
```

## Up and Downregulated Genes
To identify the regulated genes, I extract the result table, which contains, among other things, the log2FoldChanges and the adjusted p-values. I also adjust the log-fold changes based on the read coverage using the package from @apeglm.
```{r, message=FALSE}
# extract result table and add column for adjusted log fold changes
res <- results(dds, contrast = c("condition","KD","Ctrl"))
res.shrink <- lfcShrink(dds = dds, res = res, coef = 2)
res$log2FoldChange.shrink <- res.shrink$log2FoldChange
res <- as.data.frame(res)
# print only important columns
head(res[c(1,2,6)])
```

I visualize the results as an MA plot in which every data point represents a gene. The y-axis shows the log fold change, which indicates how much the expression of a gene has changed. The x-axis shows the mean of normalized counts (assigned reads), which is an indicator of the reliability of the data. All data points marked as blue represent genes with a p-value greater than 0.1. U2 related genes published by @wahl are labeled.
```{r}
res$regulated <- res$padj < 0.1
res$regulated[is.na(res$regulated)] <- FALSE
res <- mutate(res, up_down = case_when(
  regulated & log2FoldChange > 0 ~ "up",
  regulated & log2FoldChange < 0 ~ "down",
  TRUE ~ "not"
))
res$up_down <- factor(res$up_down)

# mark U2 related genes for labels
labeled_genes <- c(FUBP1="ENSG00000162613", U2AF2="ENSG00000063244", SF1="ENSG00000168066",
                   PUF60="ENSG00000179950", SPF30="ENSG00000119953", SPF31="ENSG00000126698",
                   SPF45="ENSG00000134453", CHERP="ENSG00000085872", SR140="ENSG00000241537",
                   HPrp43="ENSG00000109606", HPrp5="ENSG00000145833")
index <- match(labeled_genes, rownames(res))

# add column for labeling chosen genes
res$to.label <- ""
res$to.label[index] <- names(labeled_genes)

ggplot(res, aes(x=log10(baseMean), y=log2FoldChange, col=up_down, label=to.label))+
  geom_point(size=0.7) +
  geom_point(data=subset(res,to.label!=""),color="black") +
  geom_label_repel(max.overlaps = Inf, show.legend = FALSE)
```

In the plot above are many genes with high log fold changes and low read counts. These data points are not reliable. That is why the log fold changes have been adjusted to the read counts. In the new plot, the number of outliers with low reliability could be significantly reduced. These data suggest is that KD of FUBP1 does not lead to significantly higher expression of U2 related genes.
```{r}
ggplot(res, aes(x=log10(baseMean), y=log2FoldChange.shrink, col=up_down, label=to.label))+
  geom_point(size=0.7) +
  geom_point(data=subset(res,to.label!=""),color="black") +
  geom_label_repel(max.overlaps = Inf, show.legend = FALSE)
```

I visualize the distribution of p-values before and after adjustment, which shows the effect of multiple testing correction.
```{r}
hist(res$pvalue, main = "Distribution of raw P-Values", xlab = "P-Value")
```
 The distribution after the correction shows a lower variance.
```{r}
hist(res$padj, main = "Distribution of adjusted P-Values", xlab = "Adjusted P-Value")
```


I visualize the normalized numbers of read counts of FUBP1 and U2 related genes in the different samples. The number of read counts differs largely for the FUBP1 gene but very little or not for all U2 related genes.
```{r, message=FALSE}
counts <- counts(dds, normalized=TRUE)
colnames(counts) <- colData(dds)$condition
counts <- melt(counts) # rows all columns under each other
colnames(counts) <- c("gene.ID", "condition", "norm.count")
counts.labeled_genes <- subset(counts, gene.ID %in% labeled_genes)

labeled_genes.subst = names(labeled_genes)
names(labeled_genes.subst) <- labeled_genes

ggplot(counts.labeled_genes, aes(x=condition, y=norm.count)) + 
  geom_point() +
  scale_y_log10() +
  facet_wrap(~gene.ID, scales="free", labeller = as_labeller(labeled_genes.subst)) +
  ylim(0, 7000) +
  ylab("Normalised Counts")
```